### PR TITLE
Pyspark job to process normalization parameters

### PR DESCRIPTION
**Pull request by kaiwenw**

### CODE REVIEW

File: ml/rl/workflow/types.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    DEFAULT_MAX_QUANTILE_SIZE,
+    DEFAULT_MAX_UNIQUE_ENUM,
+    DEFAULT_NUM_SAMPLES,
+    DEFAULT_QUANTILE_K2_THRESHOLD,
+)
+
+IGNORE_SANITY_CHECK_FAILURE = True
+
+
+class TableSpec(NamedTuple):
```
Comment (kittipatv) [inline ml/rl/workflow/types.py]: Can you use dataclass instead of NamedTuple?

File: ml/rl/workflow/types.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    table: str
+    warehouse_dir: str = "spark-warehouse"
+
+
+class VarianceThreshold(NamedTuple):
```
Comment (kittipatv) [inline ml/rl/workflow/types.py]: I think we don't need this now

File: ml/rl/workflow/types.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    tablesample: Optional[float] = None
+    set_missing_value_to_zero: Optional[bool] = False
+    whitelist_features: Optional[List[int]] = None
+    assert_whitelist_feature_coverage: bool = True
+    ignore_sanity_check_failure: bool = IGNORE_SANITY_CHECK_FAILURE
+    ignore_sanity_check_task: bool = False
+    variance_threshold: VarianceThreshold = VarianceThreshold()
+    load_from_operator_id: Optional[int] = None
+    skip_sanity_check: bool = False
+    sequence_feature_id: Optional[int] = None
```
Comment (kittipatv) [inline ml/rl/workflow/types.py]: Let's leave these out since we are not implementing them (yet)

File: ml/rl/workflow/types.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    DEFAULT_NUM_SAMPLES,
+    DEFAULT_QUANTILE_K2_THRESHOLD,
+)
+
+IGNORE_SANITY_CHECK_FAILURE = True
+
+
+class TableSpec(NamedTuple):
+    table: str
+    warehouse_dir: str = "spark-warehouse"
```
Comment (kittipatv) [inline ml/rl/workflow/types.py]: This seems a bit out of place. It's only required for local spark. I'm not sure where to put it yet.

File: ml/rl/workflow/types.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    DEFAULT_MAX_UNIQUE_ENUM,
+    DEFAULT_NUM_SAMPLES,
+    DEFAULT_QUANTILE_K2_THRESHOLD,
+)
+
+IGNORE_SANITY_CHECK_FAILURE = True
+
+
+class TableSpec(NamedTuple):
+    table: str
```
Comment (kittipatv) [inline ml/rl/workflow/types.py]: Let's name this `table_url`

File: ml/rl/workflow/identify_types_flow.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+from ml.rl.preprocessing import normalization
+from ml.rl.workflow.types import PreprocessingOptions, TableSpec
+
+import pyspark
+from pyspark.sql.functions import col, collect_list, flatten
+
+
+def get_spark_session(warehouse_dir="spark-warehouse"):
+    warehouse_dir = abspath(warehouse_dir)
```
Comment (kittipatv) [inline ml/rl/workflow/identify_types_flow.py]: I think we could just leave this out. We could assume that if the user run spark locally, the warehouse location is at the directory they run the workflow from.

File: ml/rl/workflow/identify_types_flow.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        quantile_size=preprocessing_options.quantile_size,
+        quantile_k2_threshold=preprocessing_options.quantile_k2_threshold,
+        skip_box_cox=preprocessing_options.skip_box_cox,
+        skip_quantiles=preprocessing_options.skip_quantiles,
+        feature_overrides=preprocessing_options.feature_overrides,
+        whitelist_features=preprocessing_options.whitelist_features,
+        whitelist_features_assert=preprocessing_options.assert_whitelist_feature_coverage,
+    )
+
+    return normalization_processor(df)
```
Comment (kittipatv) [inline ml/rl/workflow/identify_types_flow.py]: I think we should shutdown spark session before leaving this function, e.g., with try-finally

File: ml/rl/test/workflow/test_preprocessing.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python3
```
Comment (kittipatv) [inline ml/rl/test/workflow/test_preprocessing.py]: You probably need to update ShipIt config. We wouldn't want to run this test on sandcastle

File: ml/rl/workflow/types.py (hunk 5 above)
Comment (kaiwenw) [reply ml/rl/workflow/types.py]: Right now, it's more of a table_name, not a url. Maybe table_name makes more sense?

File: ml/rl/workflow/identify_types_flow.py [hunk 9]
```diff
@@ -20,17 +20,31 @@ def get_spark_session(warehouse_dir="spark-warehouse"):
...
+        "quantile_k2_threshold": quantile_k2_threshold,
+        "skip_box_cox": skip_box_cox,
+        "skip_quantiles": skip_quantiles,
+        "feature_overrides": feature_overrides,
     }
-    whitelist_features = set(kwargs.get("whitelist_features", None) or [])
-    whitelist_features_assert = kwargs.get("whitelist_features_assert", True)
+    if whitelist_features is None:
+        whitelist_features = []
+    whitelist_features = set(whitelist_features)
```
Comment (kittipatv) [inline ml/rl/workflow/identify_types_flow.py]:
```suggestion
    whitelist_features = set(whitelist_features or [])
```

File: ml/rl/workflow/identify_types_flow.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+from pyspark.sql.functions import col, collect_list, flatten, explode
+
+from ml.rl.preprocessing import normalization
+from ml.rl.workflow.types import PreprocessingOptions, TableSpec
+
+
+def get_spark_session():
+    """ Get a spark session """
+    spark = (
+        pyspark.sql.SparkSession.builder.master("local[1]")
```
Comment (kittipatv) [inline ml/rl/workflow/identify_types_flow.py]: `"local[1]` should be an argument.

File: ml/rl/workflow/identify_types_flow.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+from typing import Dict
+from os.path import abspath
+import pyspark
+from pyspark.sql.functions import col, collect_list, flatten, explode
+
+from ml.rl.preprocessing import normalization
+from ml.rl.workflow.types import PreprocessingOptions, TableSpec
+
+
+def get_spark_session():
```
Comment (kittipatv) [inline ml/rl/workflow/identify_types_flow.py]:
```suggestion

LOCAL_MASTER = "master[1]"

def get_spark_session(master=LOCAL_MASTER):
```

Later, we can add an option to override this in the workflow

File: ml/rl/workflow/identify_types_flow.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+    # calculate fractions
+    counts_df = df.groupBy("feature_name").count()
+    frac = dict()
+    for row in counts_df.collect():
+        assert num_samples <= row["count"]
+        frac[row["feature_name"]] = num_samples / row["count"]
+
+    # perform sampling and collect them
+    df = df.sampleBy("feature_name", fractions=frac, seed=seed)
```
Comment (kittipatv) [inline ml/rl/workflow/identify_types_flow.py]: Could you add # TODO with task number and/or github issue to note that we should make this reservoir sampling?

```
# TODO(T1234, #1111): make this reservoir sampling
```

**Comment by kaiwenw** [comment]
Has non-related commits...

**Pull request closed without merging**